pulser/simulation/simulation.py

```diff
@@ -15,116 +15,302 @@
+
+        Mostly useful when dealing with multiple noise types in different
+        configurations and wanting to merge these configurations together.
+        Updates simulation parameters only when dealing with noise types that
+        weren't available in the former SimConfig.
+
+        Args:
+            config (SimConfig): SimConfig to retrieve parameters from.
+        """
+        if not isinstance(config, SimConfig):
+            raise ValueError(f"Object {config} is not a valid `SimConfig`")
+
+        old_noise_set = set(self.config.noise)
+        new_noise_set = old_noise_set.union(config.noise)
+        diff_noise_set = new_noise_set - old_noise_set
+        param_dict: dict[str, Any] = asdict(self._config)
+        del param_dict["spam_dict"]
+        del param_dict["doppler_sigma"]
+        param_dict['noise'] = tuple(new_noise_set)
+        if 'SPAM' in diff_noise_set:
+            param_dict['eta'] = config.eta
+            param_dict['epsilon'] = config.epsilon
+            param_dict['epsilon_prime'] = config.epsilon_prime
+        if 'doppler' in diff_noise_set:
+            param_dict['temperature'] = config.temperature * 1.e6
+        if 'amplitude' in diff_noise_set:
+            param_dict['laser_waist'] = config.laser_waist
+        if 'dephasing' in diff_noise_set:
+            param_dict['dephasing_prob'] = config.dephasing_prob
+        self.set_config(SimConfig(**param_dict))
+
+    def show_config(self) -> None:
+        """Shows current configuration."""
+        print(self._config)
+
+    def reset_config(self) -> None:
+        """Resets configuration to default."""
+        self.set_config(SimConfig())
+        print("Configuration has been set to default.")
+
+    def _set_param_from_config(self) -> None:
+        """Sets all relevant Simulation parameters from its SimConfig.
+
+        Called every time a new configuration is loaded in order to update
+        Simulation parameters.
+        """
+        # Noise, samples and Hamiltonian update routine
+        self._update_hamiltonian()
+        if 'dephasing' in self.config.noise and not self._collapse_ops:
+            self._init_dephasing()
+
+    def _update_hamiltonian(self) -> None:
+        """Noise, samples and Hamiltonian update routine.
+
+        Updates:
+            -Noise parameters, randomly chosen at each run
+            -New sequence samples that may have changed due to noise
+            -Sequence Hamiltonian from samples.
+        """
+        self._update_noise()
+        self._reset_samples()
+        self._extract_samples()
+        self._construct_hamiltonian()
+
+    def _update_noise(self) -> None:
+        """Updates noise random parameters.
+
+        Used at the start of each run. If SPAM isn't in chosen noises, all
+            atoms are set to be correctly prepared.
+        """
+        if 'SPAM' in self.config.noise and self.config.eta > 0:
+            self._prepare_bad_atoms()
+        elif 'SPAM' not in self.config.noise:
```

Comment:
Shouldn't be `else` instead? Don't you want to define `self._bad_atoms` when `SPAM` is in `sim.config.noise` but `eta == 0`?

Reply:
On the other hand, it doesn't make sense to be repeatedly defining `_bad_atoms` when there aren't any. When there are not preparation erros, you should define `self._bad_atoms` after you set the config instead of here